File: Leetcode/Python3/973. K Closest Points to Origin/solution.py
# https://leetcode.com/problems/k-closest-points-to-origin/


import heapq


class Solution:
    def kClosest(self, points: List[List[int]], k: int) -> List[List[int]]:
        points = sorted(points, key=lambda x: x[0]**2 + x[1]**2)
        return points[:k]


class Solution:
    def kClosest(self, points: List[List[int]], k: int) -> List[List[int]]:
        heap = []
        for x, y in points:
            heapq.heappush(heap, (x**2 + y**2, x, y))

        result = []
        for _ in range(k):
            _, x, y = heapq.heappop(heap)
            result.append([x, y])
        return result

File: Leetcode/C++/973. K Closest Points to Origin/solution.cpp
// https://leetcode.com/problems/k-closest-points-to-origin/description/

#include <algorithm>
#include <vector>

using namespace std;

// Ref: https://leetcode.com/problems/k-closest-points-to-origin/solutions/217999/java-c-python-o-n/
class Solution
{
public:
    vector<vector<int>> kClosest(vector<vector<int>> &points, int k)
    {
        /*
        Sort with compare lambda function
        The comparison function establishes the sorting criterion.
        */
        nth_element(points.begin(), points.begin() + k, points.end(), [](vector<int> &v1, vector<int> &v2)
                    { return v1[0] * v1[0] + v1[1] * v1[1] < v2[0] * v2[0] + v2[1] * v2[1]; });
        return vector<vector<int>>(points.begin(), points.begin() + k);
    }
};